Added an s-dependent IOTA lens example

cemitch99 opened the pull request:
Added a segmented IOTA lens example (Python) using 18 segments with the full s-dependence of the focusing.

The number of segments can be adjusted as a parameter.

The diagnostic output for the nonlinear invariants (H,I) needs to be updated to the corresponding values of alpha, beta, etc.  This requires that Python bindings be added for these parameters.

### Tasks

- [x] Update Testing Routines (in CMake)
- [x] Update IOTA Example Documentation (README.rst) - can also be a separate one instead of extending the existing one
- [x] Expose Python bindings for nonlinear invariants diagnostics

cemitch99 committed:
Examples for 3D space charge benchmarking

- Modified the initial beam size in the IOTA lens benchmark example.
- Added 2 benchmarks of 3D space charge for initial testing.
- Add documentation for 2 benchmarks with space charge.
- Add a benchmark example with space charge and periodic s-dependent focusing.
- Added an s-dependent example using a Kurth beam without space charge.
- Modified tolerance for IOTA lens benchmark example.
  Reduced tolerance to account for smaller initial beam size and
  improved preservation of invariants of motion.
- Modified tolerances of space charge examples to allow CI tests to
  pass when space charge is not active.

- Modified tolerance for space charge examples.
  These should fail unless space charge is turned on.

cemitch99 committed:
Update input_kurth_10nC.in

Selected numerical values for amr.n_cell, lattice.nslice, and geometry.prob_relative.

cemitch99 committed:
Add s-dependent IOTA lens example.

cemitch99 committed:
Delete examples/kurth/input_kurth_10nC.in

Not part of this PR.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add IOTA lattice with full nonlinear lens.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update run_iotalattice_sdep.py

Modified lattice to place nonlinear lens (nll) as the first element.

cemitch99 committed:
Update run_iotalattice_sdep.py

Modified initial beam moments for matching

RTSandberg commented on examples/iota_lattice/run_iotalattice_sdep.py:
```diff
@@ -0,0 +1,213 @@
+sim.lattice.append(qe3)
+lattice_after_nll.reverse()
+sim.lattice.extend(lattice_after_nll)
+sim.lattice.extend(dnll)
```
```suggestion
sim.lattice.append(dnll)
```

cemitch99 committed:
Update examples/iota_lattice/run_iotalattice_sdep.py

Co-authored-by: Ryan Sandberg <RSandberg@lbl.gov>

cemitch99 committed:
Update run_iotalattice_sdep.py

Added initial x-px and y-py correlations.

cemitch99 committed:
Add app IOTA lens example.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 commented on src/particles/diagnostics/NonlinearLensInvariants.H:
```diff
@@ -82,8 +82,8 @@ namespace impactx::diagnostics
-            amrex::ParticleReal const pyn = py*sqrt(m_beta)/m_cn + m_alpha*y;
+            amrex::ParticleReal const pxn = px*sqrt(m_beta)/m_cn + m_alpha*xn;
+            amrex::ParticleReal const pyn = py*sqrt(m_beta)/m_cn + m_alpha*yn;
 
```
Fixed a typo here.  This only affects the diagnostics when alpha is nonzero.

cemitch99 committed:
Add app input file for full IOTA lattice.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add analysis scripts.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add tests to CMakeLists.

cemitch99 committed:
Merge branch 'development' into add_iotalens_sdep

cemitch99 committed:
Update analysis_iotalens_sdep.py

Relax tolerance for (H,I) moments.

cemitch99 committed:
Update analysis_iotalattice_sdep.py

Relax tolerance of (dH,dI) slightly.

cemitch99 committed:
Update CMakeLists.txt

Comment Python-based tests until Python bindings for NL invariants are updated.

cemitch99 committed:
Update analysis_iotalattice_sdep.py

Relaxing the tolerance of dI slightly.

cemitch99 committed:
Update analysis_iotalens_sdep.py

Tune target statistics for (H,I) slightly.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update CMakeLists.txt

Turn on Python run tests, without analysis.

cemitch99 committed:
Update CMakeLists.txt

cemitch99 committed:
Update CMakeLists.txt

Use dummy analysis scripts for Python testing.

cemitch99 committed:
Add dummy Python script analysis files.

cemitch99 committed:
Update CMakeLists.txt

Fix location of analysis file.

cemitch99 committed:
Update analysis_iotalattice_sdep_dum.py

Increase tolerance.

cemitch99 committed:
Update analysis_iotalens_sdep_dum.py

Relax tolerance in dummy script.

cemitch99 committed:
Add values of alpha, beta,c,t to Python invariant calculation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update analysis_iotalens_sdep_dum.py

Tighten tolerance for Python tests (lens).

cemitch99 committed:
Update analysis_iotalattice_sdep_dum.py

Tighten tolerance Python test (lattice).

cemitch99 committed:
Update CMakeLists.txt

Eliminate unnecessary dummy analysis scripts.

cemitch99 committed:
Delete examples/iota_lattice/analysis_iotalattice_sdep_dum.py

Remove unnecessary dummy analysis scripts.

cemitch99 committed:
Delete examples/iota_lens/analysis_iotalens_sdep_dum.py

Remove unnecessary dummy analysis scriipts.

cemitch99 committed:
Update examples documentation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update analysis_iotalens_sdep.py

Pin down slight differences in initial beam generation in app vs. Python script.

cemitch99 committed:
Update analysis_iotalens_sdep.py

Tighten tolerance again

ax3l commented on examples/iota_lens/README.rst:
```diff
@@ -48,3 +48,61 @@ We run the following script to analyze correctness:
+.. _examples-iotalens-sdep:
+
+A nonlinear focusing channel based on the physical IOTA nonlinear magnet
+=========================================================================
```
```suggestion
A nonlinear focusing channel based on the physical IOTA nonlinear magnet
========================================================================
```

ax3l commented on examples/iota_lattice/README.rst:
```diff
@@ -48,3 +48,55 @@ We run the following script to analyze correctness:
+.. _examples-iotalattice-sdep:
+
+The full nonlinear lattice of the Fermilab IOTA storage ring
+=============================================================
```
```suggestion
The full nonlinear lattice of the Fermilab IOTA storage ring
============================================================
```

ax3l reviewed (approved):
Thank you, looks great ! :+1:

ax3l committed:
Tiny Formatting of Docs
